DEVX-2221: ccloud_library cleans up ccloud usage for ksqldb metadata

rspurgeon opened the pull request:
### Description 

Uses JSON processing for ksqlDB metadata in ccloud library
https://confluentinc.atlassian.net/browse/DEVX-2221

### Author Validation
- [x] cp-quickstart (cloud)

### Reviewer Tasks

- [ ] cloud-etl (if setup to validate)

andrewegel committed:
Bump Confluent to 6.1.0-SNAPSHOT, Kafka to 6.1.0-SNAPSHOT

rspurgeon committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

andrewegel committed:
Merge branch 6.0.x into master (using strategy ours)

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

andrewegel committed:
Merge branch 6.0.x into master (using strategy ours)

ybyzek committed:
Merge branch '6.0.x'

ConfluentJenkins committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Remove KAFKA vars in merge conflict

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Update connect-streams-pipeline README

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

Commit:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.x'

ybyzek committed:
Fix merge conflict

ybyzek committed:
Merge branch '6.0.x'

andrewegel committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

Commit:
Merge branch '6.0.x'

niteshmor committed:
Merge branch '6.0.x'

ybyzek committed:
Fix merge conflict

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Fix merge conflict

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

gitfrog0 committed:
Fix comparison operator typo in ccloud_library.sh (#777)

Normally, the correct credentials are found with only the resource_id, but I was trying to repurpose an existing cloud environment and things didn't work right due to the comparison being ignored.

rspurgeon committed:
Merge branch '6.0.x' into master

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
DEVX-2014: Fix init issue that surfaced in replicator demo

ybyzek committed:
Resolve merge conflicts

rspurgeon committed:
DEVX-2069: update microservices Docker image to latest official one (#786)

* DEVX-2069: Updates microservices demos to use 6.0 examples image

* DEVX-2069: Updates microservices demos to use 6.0 examples image

rspurgeon committed:
DEVX-2069: Fixes typo in microservices ccloud script output

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
GH-782 updates common makefile to use var namespace throughout (#783)

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

andrewegel committed:
Merge branch '6.0.x'

ConfluentJenkins committed:
Merge branch '6.0.x'

ConfluentJenkins committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

thomaskwscott committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

Commit:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.x'

andrewegel committed:
Merge branch '6.0.0-post' into 6.0.x

andrewegel committed:
Merge branch '6.0.x'

rspurgeon committed:
DEVX-2123: Delay reading ksqlDB topics during validation until available (#805)

* DEVX-2123: Delay reading ksqlDB topics during validation until available

* DEVX-2123: Delay reading ksqlDB topics during validation until available

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
DEVX-2070: Update datagen usage to 0.4.0 and 6.0.0 (#810)

rspurgeon committed:
DEVX-2070: Update datagen usage to 0.4.0 and 6.0.0 (#810)

andrewegel committed:
Merge branch '6.0.0-post' into 6.0.x

andrewegel committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
DEVX-2084: K8s demo updates for 6.0.0 release (#812)

* DEVX-2084: Changes related to helm / operator changes for 6.0

* DEVX-2084: Typos in K8s demos

rspurgeon committed:
DEVX-2084: K8s demo updates for 6.0.0 release (#812)

* DEVX-2084: Changes related to helm / operator changes for 6.0

* DEVX-2084: Typos in K8s demos

rspurgeon committed:
DEVX-2070: Update default datagen version in Dockerfile-local to 6.0.0 (#813)

* DEVX-2070: Update default datagen version in Dockerfile-local to 6.0.0

* DEVX-2070: Update default datagen version in Dockerfile-local to 6.0.0

rspurgeon committed:
DEVX-2070: Update default datagen version in Dockerfile-local to 6.0.0 (#813)

* DEVX-2070: Update default datagen version in Dockerfile-local to 6.0.0

* DEVX-2070: Update default datagen version in Dockerfile-local to 6.0.0

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x' into master

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

thomaskwscott committed:
Merge branch '6.0.0-post' into 6.0.x

thomaskwscott committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

andrewegel committed:
Merge branch '5.5.2-post' into 5.5.x

andrewegel committed:
Merge branch '6.0.0-post' into 6.0.x

andrewegel committed:
Merge branch '5.5.x' into 6.0.x

andrewegel committed:
Merge branch '6.0.x'

ConfluentJenkins committed:
Bump Confluent to 6.0.1-SNAPSHOT, Kafka to 6.0.1-SNAPSHOT

ConfluentJenkins committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ConfluentJenkins committed:
Bump Confluent to 5.5.3-SNAPSHOT, Kafka to 5.5.3-SNAPSHOT

ConfluentJenkins committed:
Merge branch '5.5.x' into 6.0.x

ConfluentJenkins committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

awalther28 committed:
Merge branch '6.0.0-post' into 6.0.x

awalther28 committed:
Merge branch '6.0.x' into master

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

andrewegel committed:
Merge branch '5.4.3-post' into 5.4.x

andrewegel committed:
Merge branch '5.5.2-post' into 5.5.x

andrewegel committed:
Merge branch '6.0.0-post' into 6.0.x

andrewegel committed:
Merge branch '5.4.x' into 5.5.x

andrewegel committed:
Merge branch '5.5.x' into 6.0.x

andrewegel committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x' into master

awalther28 committed:
Merge branch '6.0.0-post' into 6.0.x

awalther28 committed:
Merge branch '6.0.x' into master

ConfluentJenkins committed:
Bump Confluent to 5.4.4-SNAPSHOT, Kafka to 5.4.4-SNAPSHOT

ConfluentJenkins committed:
Merge branch '5.4.x' into 5.5.x

ConfluentJenkins committed:
Merge branch '5.5.x' into 6.0.x

ConfluentJenkins committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x' into master

elismaga committed:
ST-3461: Update to work with nano versioning (#839)

ybyzek committed:
Fix merge conflicts

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x' into master

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x' into master

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

rspurgeon committed:
DEVX-2221: ccloud_library cleans up ccloud usage for ksqldb metadata

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x' into master

ybyzek commented on utils/ccloud_library.sh:
```diff
@@ -522,30 +522,37 @@ function ccloud::validate_ccloud_config() {
   ccloud::validate_logged_in_ccloud_cli || exit 1
 
   if [[ "$ksqldb_endpoint" == "" ]]; then
     echo "ERROR: Provision a ksqlDB cluster via the Confluent Cloud UI and add the configuration parameter ksql.endpoint and ksql.basic.auth.user.info into your Confluent Cloud configuration file at $ccloud_config_file and try again."
```
`ccloud_config_file` was removed as an argument, so is the value known in this function to be able to print out the error message?

rspurgeon replied:
Sorry, can you clarify the question?  My idea with the warning is just to provide an indiciation of improper invocation of the function.

rspurgeon replied:
I understand now, sorry the GitHub UI confused me with which line you were indicating.  Thanks for catching this, will address

rspurgeon replied:
removed in latest

ybyzek commented:
Error on run:

```
[9:48:53] ~/git/examples/ccloud/ccloud-stack(DEVX-2221): ./ccloud_stack_create.sh 
This example uses real Confluent Cloud resources.
To avoid unexpected charges, carefully evaluate the cost of resources before launching the script and ensure all resources are destroyed after you are done running it.
Do you still want to run this script? [y/n] y
Do you also want to create a Confluent Cloud ksqlDB app (hourly charges may apply)? [y/n] y

Creating Confluent Cloud stack for service account demo-app-31277, ID: 132551.
Error: accepts 1 arg(s), received 29
Usage:
  ccloud kafka cluster describe <id> [flags]

Flags:
  -o, --output string   Specify the output format as "human", "json", or "yaml". (default "human")

Global Flags:
  -h, --help            Show help for this command.
  -v, --verbose count   Increase verbosity (-v for warn, -vv for info, -vvv for debug, -vvvv for trace).

Error: resource "Error" not found

Suggestions:
    Check that the resource "Error" exists.
    To list Kafka clusters, use `ccloud kafka cluster list`.
    To check schema-registry cluster info, use `ccloud schema-registry cluster describe`.
    To list KSQL clusters, use `ccloud ksql app list`.

Waiting up to 720 seconds for Confluent Cloud cluster to be ready and for credentials to propagate
.^C
```

rspurgeon commented:
> Error on run:
> 
@ybyzek can you indicate your `jq` version?

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x' into master

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

rspurgeon committed:
Merge branch '6.0.0-post' into 6.0.x

rspurgeon committed:
Merge branch '6.0.x'

ybyzek reviewed (changes requested):
IIUC this PR makes `validate_credentials_ksqldb` no longer part of the validation workflow.  This function was useful because some people try to use Kafka credentials instead of creating new ksqlDB credentials.  So it is recommended to be part of validation somehow; it can be separate from checking if the endpoint is up, but it should be part of ksqlDB validation overall

rspurgeon commented:
> IIUC this PR makes `validate_credentials_ksqldb` no longer part of the validation workflow. This function was useful because some people try to use Kafka credentials instead of creating new ksqlDB credentials. So it is recommended to be part of validation somehow; it can be separate from checking if the endpoint is up, but it should be part of ksqlDB validation overall

Thanks.  In an attempt to separate concerns and make the function more prescriptive to it's name, I did break this out.  Where the ksqlDB up check was called in external scripts, I added a additional call to the creds validation.

I'm not totally clear on the use cases you're describing that i've left out.  I've added this commit: https://github.com/confluentinc/examples/pull/855/commits/110da8d0555d65829c5ad01793ff2c8e891380ef#diff-5ebe849b3ac2695b290a288871dc904e950b16c2b0051e9d66555c4345143896R823 which adds calling the validate creds function to `function ccloud::validate_ccloud_stack_up()`, does this address your concern?  If not, would you mind adding some details on the use cases you're describing so I can add coverage for them?

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

ybyzek committed:
Merge branch '6.0.0-post' into 6.0.x

ybyzek committed:
Merge branch '6.0.x'

joel-hamill committed:
Merge branch '6.0.0-post' into 6.0.x

joel-hamill committed:
Merge branch '6.0.x'

Commit:
KC-1029 updating MRC demo with observer promotion (#857)

ybyzek commented:
@rspurgeon since a few `ccloud-stack` commits have been merged since this PR was opened, can you please rebase this to latest `6.0.0-post` and we'll give it a final run-thru before merging?

rspurgeon commented:
@ybyzek Rebased and verified with ccloud 1.20.1

ybyzek commented:
@rspurgeon FYI the current PR shows 31 modified files, including those not intended to be modified.  Maybe need rebase again or close this PR and open a new one?

rspurgeon commented:
> @rspurgeon FYI the current PR shows 31 modified files, including those not intended to be modified. Maybe need rebase again or close this PR and open a new one?

Yikes, how did I do this now... thank you for catching.  I'll figure out what I did and repost

rspurgeon commented:
closed in favor of #870
